lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1758,3 +1793,56 @@ def _relabel_configs(self) -> list:
             if self._relabel_instance
             else []
         )
+
+
+class _TransformationProvider(Object):
+    """Enables runtime transformations of promql expressions using promql-transform."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._path: PosixPath = None
+        arch = platform.processor()
+        self._arch = "amd64" if arch == "x86_64" else arch
+        self._valid_architectures = ["amd64", "arm64"]
+
+        if self._arch not in self._valid_architectures:
+            raise UnsupportedTransformArchError(self._arch)
+
+        self._fetch_transformer_path(self._arch)
+
+    def _fetch_transformer_path(self, arch: str):
+        if self._path is not None:
+            return
+
+        res = "promql-transform-{}".format(arch)
+        try:
+            self._path = self._charm.model.resources.fetch(res)
+        except NameError:
+            raise MissingTransformResourceError(res)
+
+    def apply_label_matcher(self, expr: str, additional_label_filters: dict) -> str:
+        if self._path is None:
+            logger.debug(
+                "No `promql-transform` available, skipping the application of the label matchers."
+            )
+            return expr
+
+        os.chmod(self._path, 0o755)
```

Comment:
Have to admit I do not readily recall what is 0o755. I'd [use stats](https://docs.python.org/3/library/os.html#os.chmod).

Reply:
That would require multiple modes to be combined (`IXUSR`, `IXGRP` and `IXOTH` - maybe `IWUSR` as well?) Not sure whether that would be more or less graspable than googling "linux permission 755"? I'm OK with doing either, so feel free to pick which option you like the most.

Reply:
...or even `[stat.filemode](https://docs.python.org/3/library/stat.html#stat.filemode)`.
Yes I agree.

Reply:
Would it make more sense to move the chmod into `_fetch_transformer_path`?